RFC-0008-plan-checker.md

```diff
@@ -0,0 +1,180 @@
+  converted to a Velox plan fragment, absence indicates that the plan was not converted.
+
+### Presto to Velox plan validation
+
+The Presto runtime centralizes plan validation logic into the `PlanChecker` class.  There exist three phases to this class: 
+
+* `validateIntermediatePlan`
+* `validateFinalPlan`
+* `validateFragmentedPlan`
+
+It is generally useful to allow this class to be configured with an SPI that allows for custom plan validation logic to be added.
+For example, a business may decide that a certain type is not allowed, or add a check to ensure that plans that are overly
+complicated are killed.
+
+An SPI will be added that will add more checks to the `PlanChecker` class that will allow additional checks for each of the planning
+phases.  The SPI will contain a field indicating which phase of the plan checker to be added, and a validator that will be run during
+that phase.
+
+A new endpoint to the Presto sidecar will be added that will attempt to convert a Presto plan to a Velox plan.  In the
+`presto-native-plugin` module, a new implementation of the SPI will be added which will add a check to the `PlanChecker` class
+which will call the sidecar to attempt to convert the plan.  If the conversion fails, the plan checker will fail the plan.  It will
+be added at the `validateFragmentedPlan` phase--this is because it's not until plan fragmentation occurs that we know which portions
+of the plan will be executed in the coordinator, and which will be executed in the workers.  Plan fragments that are executed in the
+coordinator, such as `COORDINATORY_ONLY` distribution types, will be skipped by the plugin.
+
+The code for the plan checker will run `PrestoToVeloxQueryPlan`, which is used by the workers to convert the Presto plan fragment
+to a Velox plan fragment.  If the conversion fails, the plan checker will fail the plan, returning with it the reason for the failure.
+
+#### Failing the plan quickly
+
+An additional code change will be made to allow the planner to execute prior to queuing.  This is so that the plan checker can
+be run before the query is queued.  This will allow the user to get feedback on the plan before the query is executed, and will
+allow the query to fail quickly if the plan is invalid.
+
+Because the queue limits concurrency, and too much concurrency during planning may require excessive resources in the coordinator,
+the plan checker will be run in a separate thread pool.  This thread pool will be configured with a maximum number of threads
+that can be run concurrently.  If the thread pool is full, then planning will wait until there is a free thread to run the planner.
+This will be configured with a new configuration parameter and session property.
+
+### EXPLAIN (TYPE NATIVE)
+
+The `EXPLAIN (TYPE NATIVE)` command will be updated to run the `PlanConverter` over all fragments which are not `COORDINATOR_ONLY`.  
```

Comment:
Should we make the endpoint take an array of fragments? We could just pass one at a time for the plan validation, but if EXPLAIN is able to gather all fragments, then it could be sent in a single request.

Reply:
One by one keeps the API simple and given it's an explain plan, not really performance sensitive (and it's O(# of fragments), and realistically we'll never have > 100 fragments).